File: 04-vector-search/python/Models/customers.py
def get_customer_addresses(addresses):
    addresses_string = ""

    for idx, address in enumerate(addresses):
        addresses_string= addresses_string + ("; " if idx > 0  else "")  \
                        + ("Address Line - " + address["addressLine1"] if address["addressLine1"]  else "")  \
                        + (" " + address["addressLine2"] if address["addressLine2"]  else "") \
                        + (", city - " + address["city"] if address["city"] else "") \
                        + (", state - " + address["state"] if address["state"] else "") \
                        + (", country - " + address["country"] if address["country"] else "") \
                        + (", zipcode - " + address["zipCode"] if address["zipCode"] else "") \
                        + (", location - " + address["location"] if address["location"] else "")  
        
    return addresses_string

def generate_customer_embedding(customer,embeddings_deployment,AzureOpenAIClient):
    if customer["type"]:
        customer["customerTypeVector"] = Embeddings.generate_embeddings (customer["type"],embeddings_deployment,AzureOpenAIClient)

    if customer["title"]:
        customer["customerTitleVector"] = Embeddings.generate_embeddings (customer["title"],embeddings_deployment,AzureOpenAIClient)

    if customer["firstName"]+" "+customer["lastName"]:
        customer["customerNameVector"] = Embeddings.generate_embeddings (customer["firstName"]+" "+customer["lastName"],embeddings_deployment,AzureOpenAIClient)

    if customer["emailAddress"]:
        customer["customerEmailAddressVector"] = Embeddings.generate_embeddings (customer["emailAddress"],embeddings_deployment,AzureOpenAIClient)

    if customer["phoneNumber"]:
        customer["customerPhoneNumberVector"] = Embeddings.generate_embeddings (customer["phoneNumber"],embeddings_deployment,AzureOpenAIClient)

    address = get_customer_addresses(customer["addresses"])
    if len(address) > 0:
        customer["customerAddressesVector"] = Embeddings.generate_embeddings (address,embeddings_deployment,AzureOpenAIClient)

    return customer

File: 04-vector-search/Solution/node.js/Collections/customers.js
// Import the Embeddings module from the SearchComponents directory
const Embeddings = require('../SearchComponents/embeddings');

// Function to convert a list of address objects into a single string
function getCustomerAddresses(addresses) {
    let addressesString = "";

    // Iterate over each address in the list
    for (let idx = 0; idx < addresses.length; idx++) {
        const address = addresses[idx];
        // Concatenate each address field into the addressesString
        addressesString += (idx > 0 ? "; " : "") +
            (address.addressLine1 ? "Address Line - " + address.addressLine1 : "") +
            (address.addressLine2 ? " " + address.addressLine2 : "") +
            (address.city ? ", city - " + address.city : "") +
            (address.state ? ", state - " + address.state : "") +
            (address.country ? ", country - " + address.country : "") +
            (address.zipCode ? ", zipcode - " + address.zipCode : "") +
            (address.location ? ", location - " + address.location : "");
    }

    // Return the concatenated string of addresses
    return addressesString;
}

// Asynchronous function to generate embeddings for various customer fields
async function generateCustomerEmbedding(customer, embeddingsDeployment, AzureOpenAIClient) {
    // If the customer has a type, generate an embedding for it
    if (customer.type) {
        customer.customerTypeVector = await Embeddings.generateEmbeddings(customer.type, embeddingsDeployment, AzureOpenAIClient);
    }

    // If the customer has a title, generate an embedding for it
    if (customer.title) {
        customer.customerTitleVector = await Embeddings.generateEmbeddings(customer.title, embeddingsDeployment, AzureOpenAIClient);
    }

    // If the customer has a first and last name, generate an embedding for it
    if (customer.firstName && customer.lastName) {
        customer.customerNameVector = await Embeddings.generateEmbeddings(customer.firstName + " " + customer.lastName, embeddingsDeployment, AzureOpenAIClient);
    }

    // If the customer has an email address, generate an embedding for it
    if (customer.emailAddress) {
        customer.customerEmailAddressVector = await Embeddings.generateEmbeddings(customer.emailAddress, embeddingsDeployment, AzureOpenAIClient);
    }

    // If the customer has a phone number, generate an embedding for it
    if (customer.phoneNumber) {
        customer.customerPhoneNumberVector = await Embeddings.generateEmbeddings(customer.phoneNumber, embeddingsDeployment, AzureOpenAIClient);
    }

    // Get the string representation of the customer's addresses
    const address = getCustomerAddresses(customer.addresses);
    // If the customer has addresses, generate an embedding for them
    if (address.length > 0) {
        customer.customerAddressesVector = await Embeddings.generateEmbeddings(address, embeddingsDeployment, AzureOpenAIClient);
    }

    // Return the customer object with the added embeddings
    return customer;
}

// Export the generateCustomerEmbedding function
module.exports.generateCustomerEmbedding = generateCustomerEmbedding;